### ISSUE TITLE
Mirror Ledger onto CashCtrl

### ISSUE THREAD

**Issue by lasuk** [macxred/cashctrl_ledger#6]
### Objective

Extend the `CashCtrlLedger` class to synchronize a general ledger entries in `pyledger` format onto a remote CashCtrl instance.

## Detailed Description

Create a method named `mirror_ledger()` within the `CashCtrlLedger` class to synchronize a set of target ledger entries, specified as a DataFrame in the `pyledger.ledger()` column schema, onto a remote CashCtrl instance. Notice that ledger entries are named 'journal entries' in CashCtrl. This method will handle the addition of new ledger entries, updates to existing entries, and optionally, the deletion of entries not present in the target DataFrame.

```python
class CashCtrlLedger(LedgerEngine):

    def mirror_ledger(self, target: pd.DataFrame, delete: bool = True):
        """
        Synchronizes ledger entries on a remote CashCtrl account with the target state provided as a DataFrame.
        Args:
            target (pd.DataFrame): DataFrame with desired target ledger entries in pyledger format.
            delete (bool, optional): If True, deletes ledger entries on the remote account that are not present in the target DataFrame.
        """
```

Field mapping between `pyledger` (left) and CashCtrl (right) includes:

- 'date' (datetime.Date) -> 'dateAdded'
- 'account' -> Look up `account['number']` via ('creditId' if amount > 0 else 'debitId')
- 'counter_account' -> Look up `account['number']` via ('debitId' if amount > 0 else 'creditId')
- 'currency' (str) -> 'currencyCode'
- 'text' (str) -> 'title'
- 'vat_code' (str | pd.NA) -> Look up `tax['name']` via 'taxId'
- 'document' (str) -> 'reference'

Ledger entries do not have unique identifiers. The logic to identify entries to add, modify or delete is more complex than for the account chart or for vat codes.

## Tasks

1. **Accessor and Mutator Methods**
   - Implement a method to retrieve the general ledger (named 'journal' in CashCtrl) in `pyledger` DataFrame format.
     ```python
     def ledger(self) -> pd.DataFrame:
         """Retrieves the account remote ledger from CashCtrl instance formatted to the pyledger schema."""
     ```
   - Develop methods for adding, updating, and deleting accounts based on the `pyledger` data.
     ```python
     def add_ledger_entry(self, ...):
         """Adds a new ledger entry to the remote CashCtrl instance."""

     def modify_ledger_entry(self, ...):
         """Updates an existing ledger entry on the remote CashCtrl instance."""

     def delete_ledger_entry(self, account: str):
         """Deletes a ledger entry from the remote CashCtrl instance."""
     ```
   - Write unit tests for these methods.

1. **Implementation of `mirror_ledger()`**
    1. Ensure remote and target data are represented by data frames with identical columns and column types. Duplicate entries are allowed.
       - Ensure the target account chart conforms to the expected schema: `target = StandaloneLedger.standardize_ledger(target).
       - Retrieve current accounts from CashCtrl. `remote = self.ledger()
   2. The algorithm needs to identify elements to keep, delete, add, modify, and perform operations using mutator methods.
        - This is conceptually an optimization task with the number of required changes as a cost function. However, we are free to take a more pragmatic approach; we do not need to guarantee optimality.
        - Our sole hard requirement is to ensure the resulting data matches (identical values, possibly arranged in a different sequence).
        - Soft requirement: limit the number of additions and deletions by modifying instead of deleting and re-creating similar but not identical entries.
        - Probably best solved by:
            1. Discarding equal elements on both sides from both data frames (caution: possible duplicates are allowed)
            2. Identifying elements that are close, modifying on the target side, discarding from both data frames.
            3. Deleting remaining elements on the target data frame.
            4. Adding remaining elements on the source data frame.
        - Maybe `difflib.get_close_matches` helps to identify similar, but not identical entries.
   - Extend unit tests to cover the `mirror_ledger()` method.
       - include scenarios where the target DataFrame is empty, contains NA `vat_code` and `counter_account`.
       - include scenarios where  `vat_code` is not defined
       - include scenarios where  `account`, or `counter_accounts` are not defined in the account chart.

1. **Submit Pull Request**
   - Submit the changes for review, assigning the PR to Lukas.

### Test Data

1. Generate some initial test data in the CashCtrl GUI.
1. Lukas can provide a more challenging set of test data as work progresses.

**Comment by lasuk** [comment]
### Data Structure

Note that ledger data frames in pyledger can mix both single and collective postings.

- **Single postings** are financial transactions where an amount is debited from a single account and credited to a single account. These transactions occupy a single row in the ledger data frame. A minimal definition includes the 'date', 'currency', 'amount', a credit 'account', a 'counter account' to debit, and a 'text' explanation. Optional fields include 'vat_code' and 'document'.

- **Collective postings** are financial transactions where one or more accounts are debited and credited in a single transaction on the same date. The amounts credited minus the amounts debited must sum to zero in the base currency of the accounting system. These transactions occupy multiple rows in the ledger data frame. To denote multiple lines forming a single transaction, the date is written only on the first row, with the subsequent rows of the same transaction having NA in the date column.

In summary: subsequent rows with a missing (NA) date form a single transaction together with the last row that has a non-missing date.

**Comment by lasuk** [comment]
Assigning ledger data to single transactions is implemented in `LedgerEngine. standardize_ledger`. 
```
df['id'] = df['date'].notna().cumsum().astype(pd.StringDtype())
```
See [pyledger/ledger_engine.py#L621](https://github.com/macxred/pyledger/blob/bb5858f8b1/pyledger/ledger_engine.py#L621)

**Comment by AlexTheWizardL** [comment]
Reopened for the mirror function
